AppController/lib/haproxy.rb

```diff
@@ -84,47 +85,65 @@ def self.app_listen_port(app_number)
     START_PORT + app_number
   end
 
-  # Create the config file for Datastore Server applications.
-  def self.create_datastore_server_config(my_ip, listen_port, table)
-    self.create_app_config(my_ip, my_ip, listen_port, 
-      DatastoreServer.get_server_ports(table), DatastoreServer::NAME)
+  # Create the config file for UserAppServer.
+  def self.create_ua_server_config(servers, my_ip, listen_port)
+    # We reach out to UserAppServers on the DB nodes.
+    # The port is fixed.
+    ports = []
+    servers.each{ |server|
+      ports << UserAppClient::SERVER_PORT
+    }
+    self.create_app_config(servers, my_ip, listen_port, ports,
+      UserAppClient::NAME)
   end
 
-  # Create the configuration file for the AppDashboard application.
-  def self.create_app_load_balancer_config(my_public_ip, my_private_ip,
-    listen_port)
-    self.create_app_config(my_public_ip, my_private_ip, listen_port,
-      AppDashboard::SERVER_PORTS, AppDashboard::APP_NAME)
+  # Create the config file for Datastore Server.
+  def self.create_datastore_server_config(my_ip, listen_port, table)
+    # For the Datastore servers we have a list of local ports the servers
+    # are listening to, and we need to create the list of local IPs.
+    ips = []
+    DatastoreServer.get_server_ports(table).each { |port|
+      ips << my_ip
+    }
+    self.create_app_config(ips, my_ip, listen_port,
+      DatastoreServer.get_server_ports(table), DatastoreServer::NAME)
   end
 
-  # A generic function for creating haproxy config files used by appscale services
-  def self.create_app_config(my_public_ip, my_private_ip, listen_port, 
+  # A generic function for creating HAProxy config files used by AppScale services.
+  #
+  # Arguments:
+  #   server_ips  : list of server IPs
+  #   listen_ip   : the IP HAProxy should listen for
+  #   listen_port : the port to listen to
+  #   server_ports: list of server ports, corresponding to server_ips
+  #   name        : the name of the server
+  def self.create_app_config(server_ips, my_private_ip, listen_port,
     server_ports, name)
     servers = []
     server_ports.each_with_index do |port, index|
-      servers << HAProxy.server_config(name, index, "#{my_private_ip}:#{port}")
+      servers << HAProxy.server_config(name, index, "#{server_ips[index]}:#{port}")
```

Comment:
It's a bit counter-intuitive to me to have arguments that are grouped by index. What do you think about a single list of hashes (eg with the keys 'ip' and 'port')?